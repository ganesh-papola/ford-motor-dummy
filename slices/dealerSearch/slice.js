import { createSlice } from '@reduxjs/toolkit';
import { reducers } from './reducers';
import {
  autoAllocateDealerFromBrowserLocation,
  allocateDealer,
  getSuggestedDealers,
} from './thunks';
import initialState from './initialState';
import {
  updateDeliveryOptionsSuccess,
  updateReservationDeliveryData,
  updateSelectedDeliveryMethod,
} from '@Ducks/deliveryData/actions';
import {
  createCartTry,
  getCartSuccess,
  orderToCartSuccess,
} from '@Ducks/sap/actions';
import { DELIVERY_CARDS_TYPES } from '@Constants/main';
import {
  getEstimatedGroupByFromGetCartResponseBody,
  isNonEmptyString,
} from '@Slices/dealerSearch/utilities';

export const validDeliveryMethodTypeValues = Object.values(
  DELIVERY_CARDS_TYPES
);

export const slice = createSlice({
  name: 'dealerSearch',
  initialState,
  reducers,
  extraReducers: (builder) => {
    builder.addCase(autoAllocateDealerFromBrowserLocation.pending, (state) => {
      state.dataRequests.autoAllocateDealerFromBrowserLocation.isLoading = true;
      state.dataRequests.autoAllocateDealerFromBrowserLocation.error = null;
    });

    builder.addCase(
      autoAllocateDealerFromBrowserLocation.fulfilled,
      (state, { payload }) => {
        state.dataRequests.autoAllocateDealerFromBrowserLocation.isLoading = false;
        state.dataRequests.autoAllocateDealerFromBrowserLocation.error = null;
        state.selectedDealer.distance = payload?.distanceToDealer ?? null;
      }
    );

    builder.addCase(
      autoAllocateDealerFromBrowserLocation.rejected,
      (state, { error }) => {
        state.dataRequests.autoAllocateDealerFromBrowserLocation.isLoading = false;
        state.dataRequests.autoAllocateDealerFromBrowserLocation.error = error;
      }
    );

    builder.addCase(allocateDealer.pending, (state) => {
      state.dataRequests.allocateDealer.isLoading = true;
      state.dataRequests.allocateDealer.error = null;
    });

    builder.addCase(allocateDealer.fulfilled, (state, { payload }) => {
      state.dataRequests.allocateDealer.isLoading = false;
      state.selectedDealer.distance = payload?.distanceToDealer ?? null;
    });

    builder.addCase(allocateDealer.rejected, (state, { error }) => {
      state.dataRequests.allocateDealer.isLoading = false;
      state.dataRequests.allocateDealer.error = error;
    });

    builder.addCase(getSuggestedDealers.pending, (state) => {
      state.dataRequests.getSuggestedDealers.error = null;
      state.dataRequests.getSuggestedDealers.isLoading = true;
    });

    builder.addCase(getSuggestedDealers.fulfilled, (state, { payload }) => {
      state.dataRequests.getSuggestedDealers.isLoading = false;
      state.suggestedDealers = payload.dealerSearchResults.vendors;
      state.dealerLeadTimeGroupMatched =
        payload.dealerSearchResults.dealerLeadTimeGroupMatched;
      state.searchParams.delivery.latitude =
        payload.deliveryAddress.position[0];
      state.searchParams.delivery.longitude =
        payload.deliveryAddress.position[1];
    });

    builder.addCase(getSuggestedDealers.rejected, (state, { error }) => {
      state.dataRequests.getSuggestedDealers.error = error;
      state.dataRequests.getSuggestedDealers.isLoading = false;
      state.suggestedDealers = [];
    });

    builder.addCase(createCartTry().type, (state, { payload }) => {
      state.searchParams.groupBy = payload?.groupBy ?? null;
    });

    builder.addCase(getCartSuccess().type, (state, { payload }) => {
      const hasInventory = payload?.entries?.[0]?.hasInventory ?? null;
      if (hasInventory !== null) {
        state.searchParams.hasInventory = hasInventory;
      }

      // TODO: US3150880: revisit showroom to cart browser storage api in order to remove this

      // assign groupBy from getCart response if we receive it
      const groupByFromGetCart = payload?.groupBy ?? null;
      const groupByFromGetCartIsValid = isNonEmptyString(groupByFromGetCart);
      if (groupByFromGetCartIsValid) {
        state.searchParams.groupBy = groupByFromGetCart;
        return;
      }

      // otherwise assign estimated groupBy if we don't already have it in state
      const groupByInState = state.searchParams.groupBy;
      const isValidGroupBy = isNonEmptyString(groupByInState);
      if (isValidGroupBy) return;

      const estimatedGroupBy = getEstimatedGroupByFromGetCartResponseBody(
        payload
      );
      const isValidEstimatedGroupBy = isNonEmptyString(estimatedGroupBy);
      if (!isValidEstimatedGroupBy) return;

      state.searchParams.groupBy = estimatedGroupBy;
    });

    builder.addMatcher(
      ({ type }) =>
        [orderToCartSuccess().type, getCartSuccess().type].includes(type),
      (state, { payload }) => {
        const additionalFilters =
          payload?.entries?.[0]?.dealer?.additionalFilters ?? null;
        if (additionalFilters === null) return;

        state.searchParams.additionalFilters = additionalFilters;
      }
    );

    // TODO: US3150880: write a helper function to perform this repeated array manipulation
    builder.addMatcher(
      ({ type }) => {
        return [getCartSuccess, updateDeliveryOptionsSuccess]
          .map((actionCreator) => actionCreator().type)
          .includes(type);
      },
      (state, { payload }) => {
        const deliveryMethodType =
          payload?.deliveryData?.deliveryMethodSelected?.type ?? null;
        const isValidDeliveryMethodType =
          deliveryMethodType !== null &&
          validDeliveryMethodTypeValues.includes(deliveryMethodType);

        if (!isValidDeliveryMethodType) return;

        state.searchParams.delivery.methodType = deliveryMethodType;
      }
    );

    builder.addMatcher(
      ({ type }) => {
        return [updateSelectedDeliveryMethod, updateReservationDeliveryData]
          .map((actionCreator) => actionCreator().type)
          .includes(type);
      },
      (state, { payload }) => {
        const deliveryMethodType = payload?.deliveryMethodSelected ?? null;
        const isValidDeliveryMethodType =
          deliveryMethodType !== null &&
          validDeliveryMethodTypeValues.includes(deliveryMethodType);

        if (!isValidDeliveryMethodType) return;

        state.searchParams.delivery.methodType = deliveryMethodType;
      }
    );
  },
});

// TODO: US3150880: Remove updateDeliveryOptionsSuccess from this slice

export const { actions, reducer } = slice;
