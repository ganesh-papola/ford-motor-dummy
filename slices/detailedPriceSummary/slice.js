import initialState from './initialState';
import { createSlice } from '@reduxjs/toolkit';
import * as sapTypes from '@Ducks/sap/types';

export const slice = createSlice({
  name: 'detailedPriceSummary',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(
      sapTypes.FETCH_RESERVATION_DETAILS_SUCCESS,
      (state, action) => {
        state.dealerizedPrice.totalVehiclePriceWithSavings =
          action?.payload?.data?.entries[0]?.detailedPriceSummary
            ?.dealerizedPrice?.totalVehiclePriceWithSavings ?? null;

        state.nonDealerizedPrice.totalPriceWithoutIncentives.grossRetailWithVehicleRegistrationTax =
          action?.payload?.data?.entries[0]?.detailedPriceSummary
            ?.nonDealerizedPrice?.totalPriceWithoutIncentives
            ?.grossRetailWithVehicleRegistrationTax ?? null;
      }
    );
    builder.addCase(sapTypes.GET_CART_SUCCESS, (state, action) => {
      state.dealerizedPrice.totalVehiclePriceWithSavings =
        action?.payload?.entries[0]?.detailedPriceSummary?.dealerizedPrice
          ?.totalVehiclePriceWithSavings ?? null;

      const basePriceArray =
        action?.payload?.entries[0]?.detailedPriceSummary?.nonDealerizedPrice
          ?.basePrice?.prices;

      const grossRetailWithVehicleRegistrationTaxBasePrice = getValueForSpecifiedType(
        basePriceArray,
        'grossRetailWithVehicleRegistrationTax'
      );

      state.nonDealerizedPrice.basePrice.grossRetailWithVehicleRegistrationTax = grossRetailWithVehicleRegistrationTaxBasePrice;

      const totalPriceWithoutIncentivesArray =
        action?.payload?.entries[0]?.detailedPriceSummary?.nonDealerizedPrice
          ?.totalPriceWithoutIncentives?.prices;

      const grossRetailWithVehicleRegistrationTaxTotalPriceWithoutIncentives = getValueForSpecifiedType(
        totalPriceWithoutIncentivesArray,
        'grossRetailWithVehicleRegistrationTax'
      );

      state.nonDealerizedPrice.totalPriceWithoutIncentives.grossRetailWithVehicleRegistrationTax = grossRetailWithVehicleRegistrationTaxTotalPriceWithoutIncentives;
    });
  },
});

function getValueForSpecifiedType(array = [], specifiedType = '') {
  const filteredArray = array.filter(({ type }) => type === specifiedType);
  return filteredArray[0]?.value ?? null;
}

export const { reducer } = slice;
