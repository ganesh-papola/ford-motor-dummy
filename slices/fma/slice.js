import { createSlice } from '@reduxjs/toolkit';
import { reducers } from './reducers';

export const name = 'fma';

export const initialState = {
  isAuthenticated: undefined,
  overrideFmaRequired: false,
  access_token: null,
  isLoading: false,
  guestGUID: null,
};

export const slice = createSlice({
  name,
  initialState,
  reducers,
});

export const actions = slice.actions;
export const reducer = slice.reducer;
