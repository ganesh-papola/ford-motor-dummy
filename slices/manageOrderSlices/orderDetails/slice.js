import { createSlice } from '@reduxjs/toolkit';
import { reducers } from './reducers';
import { fetchManageOrderDetails } from './thunks';
import initialState from './initialState';

export const slice = createSlice({
  name: 'manageOrderData',
  initialState,
  reducers,
  extraReducers: (builder) => {
    builder.addCase(fetchManageOrderDetails.pending, (state) => {
      state.isLoading = true;
      state.hasFailed = false;
      state.hasSucceeded = false;
    });

    builder.addCase(fetchManageOrderDetails.fulfilled, (state, action) => {
      state.isLoading = false;
      state.hasFailed = false;
      state.hasSucceeded = true;
      state.hasSucceededWithValidData = true;
      state.data = action.payload;
    });

    builder.addCase(fetchManageOrderDetails.rejected, (state) => {
      state.isLoading = false;
      state.hasFailed = true;
      state.hasSucceeded = false;
    });
  },
});

export const { actions, reducer } = slice;
