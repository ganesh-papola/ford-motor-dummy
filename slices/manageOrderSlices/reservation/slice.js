import { createSlice } from '@reduxjs/toolkit';
import { reducers } from './reducers';
import { fetchReservationDetails } from './thunks';
import initialState from './initialState';

export const slice = createSlice({
  name: 'reservation',
  initialState,
  reducers,
  extraReducers: (builder) => {
    builder.addCase(fetchReservationDetails.pending, (state) => {
      state.isLoading = true;
      state.hasFailed = false;
      state.hasSucceeded = false;
    });

    builder.addCase(fetchReservationDetails.fulfilled, (state) => {
      state.isLoading = false;
      state.hasFailed = false;
      state.hasSucceeded = true;
      state.hasSucceededWithValidData = true;
    });

    builder.addCase(fetchReservationDetails.rejected, (state) => {
      state.isLoading = false;
      state.hasFailed = true;
      state.hasSucceeded = false;
    });
  },
});

export const { actions, reducer } = slice;
