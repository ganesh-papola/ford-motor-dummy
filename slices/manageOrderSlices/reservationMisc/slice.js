import initialState from './initialState';
import { createSlice } from '@reduxjs/toolkit';
import * as sapTypes from '@Ducks/sap/types';

export const slice = createSlice({
  name: 'reservationMisc',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(
      sapTypes.FETCH_RESERVATION_DETAILS_SUCCESS,
      (state, action) => {
        state.finalPayment = action?.payload?.data?.finalPayment ?? null;
      }
    );
  },
});

export const { reducer } = slice;
