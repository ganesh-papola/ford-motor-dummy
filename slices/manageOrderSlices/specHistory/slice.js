import initialState from './initialState';
import { createSlice } from '@reduxjs/toolkit';
import { reducers } from './reducers';
import { fetchAmendSpecHistory } from '@Slices/manageOrderSlices/specHistory/thunk';

export const slice = createSlice({
  name: 'amendSpecHistory',
  initialState,
  reducers,
  extraReducers: (builder) => {
    builder.addCase(fetchAmendSpecHistory.pending, (state) => {
      state.isLoading = true;
      state.hasFailed = false;
      state.hasSucceeded = false;
    });
    builder.addCase(fetchAmendSpecHistory.fulfilled, (state, action) => {
      state.isLoading = false;
      state.hasFailed = false;
      state.hasSucceeded = true;
      state.hasSucceededWithValidData = true;
      state.data = action.payload;
    });
    builder.addCase(fetchAmendSpecHistory.rejected, (state) => {
      state.isLoading = false;
      state.hasFailed = true;
      state.hasSucceeded = false;
    });
  },
});

export const { actions, reducer } = slice;
