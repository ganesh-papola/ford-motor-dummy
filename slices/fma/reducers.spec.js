import { reducer } from './slice';
import { actions } from './slice';
import { initialState } from './slice';

describe('fma reducers', () => {
  it('should load initialState', () => {
    expect(reducer(initialState, {})).toMatchSnapshot();
  });

  describe('fmaIsAuthenticated', () => {
    it('should set isAuthenticated to payload', () => {
      const state = reducer(
        initialState,
        actions.authenticationStateChanged(true)
      );
      expect(state.isAuthenticated).toBe(true);
    });
  });

  describe('paymentPlanFail', () => {
    it('should set overrideFmaRequired to payload', () => {
      const state = reducer(initialState, actions.setOverrideFmaRequired(true));
      expect(state.overrideFmaRequired).toBe(true);
    });
  });
});
