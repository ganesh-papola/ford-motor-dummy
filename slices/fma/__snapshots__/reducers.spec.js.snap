// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fma reducers should load initialState 1`] = `
Object {
  "access_token": null,
  "guestGUID": null,
  "isAuthenticated": undefined,
  "isLoading": false,
  "overrideFmaRequired": false,
}
`;
